pallet/staking/src/lib.rs

```diff
@@ -413,64 +417,42 @@ pub mod pallet {
 				return Ok(());
 			}
 
+			let mut acc = <AccumulateUnstake<T>>::get().saturating_add(ring_amount);
+
 			<Ledgers<T>>::try_mutate(&who, |l| {
 				let l = l.as_mut().ok_or(<Error<T>>::NotStaker)?;
 
 				if ring_amount != 0 {
-					Self::unstake_ring(&mut l.staked_ring, &mut l.unstaking_ring, ring_amount)?;
-				}
+					l.ring = l
+						.ring
+						.checked_sub(ring_amount)
+						.ok_or("[pallet::staking] `u128` must not be overflowed; qed")?;
 
-				for d in deposits {
-					Self::unstake_deposit(l, d)?;
+					<T as Config>::Ring::unstake(&who, ring_amount)?;
 				}
 
-				DispatchResult::Ok(())
-			})?;
-
-			Ok(())
-		}
-
-		/// Cancel the `unstake` operation.
-		///
-		/// Re-stake the unstaking assets immediately.
-		#[pallet::call_index(2)]
-		#[pallet::weight(<T as Config>::WeightInfo::restake(deposits.len() as _))]
-		pub fn restake(
-			origin: OriginFor<T>,
-			ring_amount: Balance,
-			deposits: Vec<DepositId<T>>,
-		) -> DispatchResult {
-			let who = ensure_signed(origin)?;
-
-			if ring_amount == 0 && deposits.is_empty() {
-				return Ok(());
-			}
-
-			<Ledgers<T>>::try_mutate(&who, |l| {
-				let l = l.as_mut().ok_or(<Error<T>>::NotStaker)?;
+				for d in deposits {
+					acc = acc.saturating_add(T::Deposit::amount(&who, d).unwrap_or_default());
 
-				if ring_amount != 0 {
-					Self::restake_ring(&mut l.staked_ring, &mut l.unstaking_ring, ring_amount)?;
-				}
+					l.deposits.remove(
+						l.deposits
+							.iter()
+							.position(|d_| d_ == &d)
+							.ok_or(<Error<T>>::DepositNotFound)?,
+					);
 
-				for d in deposits {
-					Self::restake_deposit(l, d)?;
+					T::Deposit::unstake(&who, d)?;
 				}
 
 				DispatchResult::Ok(())
 			})?;
 
-			Ok(())
-		}
-
-		/// Claim the stakes from the pallet/contract account.
-		#[pallet::call_index(3)]
-		#[pallet::weight(<T as Config>::WeightInfo::claim())]
-		pub fn claim(origin: OriginFor<T>) -> DispatchResult {
-			let who = ensure_signed(origin)?;
+			if acc <= <MaxUnstakeRing<T>>::get() {
+				<AccumulateUnstake<T>>::put(acc);
+			} else {
+				Err(<Error<T>>::ExceedMaxUnstakeAmount)?;
```

Comment:
Just want to make sure:

This is transactional right? The state changes will revert when meet `Err` here?

Reply:
Yep, check the *assert no op* tests.

Reply:
OK, cool, I forgot transnational is default now.